README.md


**New in 2.5.0:** Allow reference numbers to be set in brackets ([Issue #57](https://github.com/tomduck/pandoc-eqnos/issues/57)).

**New in 2.4.0:** Updated to work with pandoc 2.11.

[more...](#whats-new)


pandoc-eqnos 2.5.0
==================

*pandoc-eqnos* is a [pandoc] filter for numbering equations and their references when converting markdown to other formats.  It is part of the [pandoc-xnos] filter suite.  LaTeX/pdf, html, and epub output have native support.  Native support for docx output is a work in progress.

Demonstration: Processing [demo3.md] with pandoc + pandoc-eqnos gives numbered equations and references in [pdf][pdf3], [tex][tex3], [html][html3], [epub][epub3], [docx][docx3] and other formats (including beamer slideshows).

This version of pandoc-eqnos was tested using pandoc 1.15.2 - 2.11.1.1,<sup>[1](#footnote1)</sup> and may be used with linux, macOS, and Windows. Bug reports and feature requests may be posted on the project's [Issues tracker].  If you find pandoc-eqnos useful, then please kindly give it a star [on GitHub].

See also: [pandoc-fignos], [pandoc-tablenos], [pandoc-secnos] \
Other filters: [pandoc-comments], [pandoc-latex-extensions]

[pandoc]: http://pandoc.org/
[pandoc-xnos]: https://github.com/tomduck/pandoc-xnos
[Issues tracker]: https://github.com/tomduck/pandoc-eqnos/issues
[on GitHub]:  https://github.com/tomduck/pandoc-eqnos
[pandoc-fignos]: https://github.com/tomduck/pandoc-fignos
[pandoc-tablenos]: https://github.com/tomduck/pandoc-tablenos
[pandoc-secnos]: https://github.com/tomduck/pandoc-secnos
[pandoc-comments]: https://github.com/tomduck/pandoc-comments
[pandoc-latex-extensions]: https://github.com/tomduck/pandoc-latex-extensions


Contents
--------

 1. [Installation](#installation)
 2. [Usage](#usage)
 3. [Markdown Syntax](#markdown-syntax)
 4. [Customization](#customization)
 5. [Technical Details](#technical-details)
 6. [Getting Help](#getting-help)
 7. [Development](#development)
 8. [What's New](#whats-new)


Installation
------------

Pandoc-eqnos requires [python].  It is easily installed -- see [here](https://realpython.com/installing-python/).<sup>[2](#footnote2)</sup>  Either python 2.7 or 3.x will do.

Pandoc-eqnos may be installed using the shell command

    pip install pandoc-eqnos --user

and upgraded by appending `--upgrade` to the above command.  Pip is a program that downloads and installs software from the Python Package Index, [PyPI].  It normally comes installed with a python distribution.<sup>[3](#footnote3)</sup>

Instructions for installing from source are given in [DEVELOPERS.md].

[python]: https://www.python.org/
[PyPI]: https://pypi.python.org/pypi
[DEVELOPERS.md]: DEVELOPERS.md


Usage
-----

Pandoc-eqnos is activated by using the

    --filter pandoc-eqnos

option with pandoc.  Alternatively, use

    --filter pandoc-xnos

to activate all of the filters in the [pandoc-xnos] suite (if installed).

Any use of `--filter pandoc-citeproc` or `--bibliography=FILE` should come *after* the `pandoc-eqnos` or `pandoc-xnos` filter calls.


Markdown Syntax
---------------

The cross-referencing syntax used by pandoc-eqnos was developed in [pandoc Issue #813] -- see [this post] by [@scaramouche1].

To mark an equation for numbering, add an identifier to its attributes:

    $$ y = mx + b $$ {#eq:id}

The prefix `#eq:` is required. `id` should be replaced with a unique string composed of letters, numbers, dashes and underscores.  If `id` is omitted then the equation will be numbered but unreferenceable.

To reference the equation, use

    @eq:id

or

    {@eq:id}

Curly braces protect a reference and are stripped from the output.

Demonstration: Processing [demo.md] with pandoc + pandoc-eqnos gives numbered equations and references in [pdf], [tex], [html], [epub], [docx] and other formats.

[pandoc Issue #813]: https://github.com/jgm/pandoc/issues/813
[this post]: https://github.com/jgm/pandoc/issues/813#issuecomment-70423503
[@scaramouche1]: https://github.com/scaramouche1
[demo.md]: https://raw.githubusercontent.com/tomduck/pandoc-eqnos/master/demos/demo.md
[pdf]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo.pdf
[tex]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo.tex
[html]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo.html
[epub]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo.epub
[docx]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo.docx


### Clever References ###

Writing markdown like

    See eq. @eq:id.

seems a bit redundant.  Pandoc-eqnos supports "clever references" via single-character modifiers in front of a reference.  You can write

     See +@eq:id.

to have the reference name (i.e., "eq.") automatically generated.  The above form is used mid-sentence; at the beginning of a sentence, use

     *@eq:id

instead.  If clever references are enabled by default (see [Customization](#customization), below), then users may disable it for a given reference using<sup>[4](#footnote4)</sup>

    !@eq:id

Demonstration: Processing [demo2.md] with pandoc + pandoc-eqnos gives numbered equations and references in [pdf][pdf2], [tex][tex2], [html][html2], [epub][epub2], [docx][docx2] and other formats.

Note: When using `*@eq:id` and emphasis (e.g., `*italics*`) in the same sentence, the `*` in the clever reference must be backslash-escaped; i.e., `\*@eq:id`.

[demo2.md]: https://raw.githubusercontent.com/tomduck/pandoc-eqnos/master/demos/demo2.md
[pdf2]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo2.pdf
[tex2]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo2.tex
[html2]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo2.html
[epub2]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo2.epub
[docx2]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo2.docx


### Tagged Equations ###

The equation number may be overridden by placing a tag in the equation's attributes block:

    $$ y = mx + b $$ {#eq:id tag="B.1"}

The tag may be arbitrary text, or an inline equation such as `$\mathrm{B.1'}$`.  Mixtures of the two are not currently supported.


### Disabling Links ###

To disable a link on a reference, set `nolink=True` in the reference's attributes:

    @eq:id{nolink=True}


Customization
-------------

Pandoc-eqnos may be customized by setting variables in the [metadata block] or on the command line (using `-M KEY=VAL`).  The following variables are supported:

  * `eqnos-warning-level` or `xnos-warning-level` - Set to `0` for
    no warnings, `1` for critical warnings, or `2` (default) for
    all warnings.  Warning level 2 should be used when
    troubleshooting.

  * `eqnos-cleveref` or `xnos-cleveref` - Set to `True` to assume "+"
    clever references by default;

  * `xnos-capitalise` - Capitalises the names of "+" clever
    references (e.g., change from "fig." to "Fig.");

  * `eqnos-plus-name` - Sets the name of a "+" clever reference 
    (e.g., change it from "eq." to "equation"). Settings here take
    precedence over `xnos-capitalise`;

  * `eqnos-star-name` - Sets the name of a "*" clever reference 
    (e.g., change it from "Equation" to "Eq.");

  * `eqnos-eqref` - Set to `True` to use AMS-style equation references
     (i.e., equation numbers set in brackets); and

  * `eqnos-number-by-section` or `xnos-number-by-section` - Set to
    `True` to number equations by section (e.g., Eq. 1.1, 1.2, etc in
    Section 1, and Eq. 2.1, 2.2, etc in Section 2).  For LaTeX/pdf,
    html, and epub output, this feature should be used together with
    pandoc's `--number-sections`
    [option](https://pandoc.org/MANUAL.html#option--number-sections)
    enabled.  For docx, use [docx custom styles] instead.

    This option should not be set for numbering by chapter in
    LaTeX/pdf book document classes.

  * `xnos-number-offset` - Set to an integer to offset the section
    numbers when numbering equations by section.  For html and epub
    output, this feature should be used together with pandoc's
    `--number-offset`
    [option](https://pandoc.org/MANUAL.html#option--number-sections)
    set to the same integer value.  For LaTeX/pdf, this option
    offsets the actual section numbers as required.

  * `eqnos-default-env` - Name of the default LaTeX environment
    (default: 'equation').  See [Environments](#environments), below.

Note that variables beginning with `eqnos-` apply to only pandoc-eqnos, whereas variables beginning with `xnos-` apply to all of the pandoc-fignos/eqnos/tablenos/secnos.

Demonstration: Processing [demo3.md] with pandoc + pandoc-eqnos gives numbered equations and references in [pdf][pdf3], [tex][tex3], [html][html3], [epub][epub3], [docx][docx3] and other formats.

[metadata block]: http://pandoc.org/README.html#extension-yaml_metadata_block
[docx custom styles]: https://pandoc.org/MANUAL.html#custom-styles
[demo3.md]: https://raw.githubusercontent.com/tomduck/pandoc-eqnos/master/demos/demo3.md
[pdf3]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo3.pdf
[tex3]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo3.tex
[html3]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo3.html
[epub3]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo3.epub
[docx3]: https://raw.githack.com/tomduck/pandoc-eqnos/demos/demo3.docx


### Environments ###

The default LaTeX environment may be overridden by adding an `env` attribute.  For example:

    $$ y = mx + b $$ {#eq:id env=multline}

The `env` attribute must be a valid amsmath environment.
If the attribute value is of the form `X.Y`, `X` will be used
as the name of the amsmath environment and `Y` will be used as an extra argument for the environment (e.g. the `alignat` environment expects an argument for the number of equation columns).  This customization only affects LaTeX/PDF output only.


Technical Details
-----------------

### LaTeX/pdf Output ###

During processing, pandoc-eqnos inserts the packages and supporting LaTeX it needs into the `header-includes` metadata field.  To see what is inserted, set the `eqnos-warning-level` meta variable to `2`.  Note that any use of pandoc's `--include-in-header` option [overrides](https://github.com/jgm/pandoc/issues/3139) all `header-includes`.

An example reference in LaTeX looks like

~~~latex
See \cref{eq:polynomial}.
~~~

An example equation looks like

~~~latex
\begin{equation}
  y = \sum_{n=0}^{\infty} a_n x^n
  \label{eq:polynomial}
\end{equation}
~~~

Other details:


ChangeLog
pandoc-eqnos 2.5.0 (2020-12-05)

    * Allow reference numbers to be set in brackets (Issue #57).


pandoc-eqnos 2.4.0 (2020-11-10)

    * Require pandoc-xnos 2.5.0 (updated to work with pandoc 2.11).


pandoc-eqnos 2.3.0 (2020-08-08)

    * Allow LaTeX environment customization (Pull Request #44).


pandoc-eqnos 2.2.3 (2020-08-05)

    * Fixed XHTML Transitional validation error (Issue #50).


pandoc-eqnos 2.2.1 (2020-07-31)

    * Require pandoc-xnos 2.4.2 (bug fix release).


pandoc-eqnos 2.2.0 (2020-07-30)

    * Updated to work with pandoc 2.10.
    * Fixed "local variable 'ret' referenced before assignment"
      error (Issue #43).
    * Fixed equation number reset problem for html, docx (Issue #49).
    * Added html4 as an output target (Issue #50).


pandoc-eqnos 2.1.1 (2019-09-26)

    * Warns about duplicate reference targets.


pandoc-eqnos 2.0.0 (2019-09-17)

    * A warnings system has been added.
    * Meta variable names have been updated.
    * TeX codes have been completely overhauled.
    * Html/epub equations are now enclosed in a `<div>`.
    * Epub support is generally improved.
    * Support for disabling links on selected references
      (pandoc-fignos Issue #54).
    * Enable section number offsets in equation numbers
      (pandoc-fignos Issue #47).
    * Better handling of missing references.
    * Changed `eqnos-number-sections` and `xnos-number-sections`
      options to `eqnos-number-by-section` and
      `xnos-number-by-section`


pandoc-eqnos 1.4.4 (2019-06-15)

    * For epub output, place equation number in own span (Issue #39).


pandoc-eqnos 1.4.3 (2019-05-17)

    * Improved epub output and refreshed epub demos.


pandoc-eqnos 1.4.2 (2019-05-16)

    * Fixed links in epub output (Issue #37).
    * Bug fix for equation numbering by section for docx
      (pandoc-fignos Issue #58).
    * `pandoc-eqnos --version` now gives the verson.


pandoc-eqnos 1.4.1 (2019-05-16)

    * Restored eqnos-eqref functionality (Issue #35).


pandoc-eqnos 1.4.0 (2019-05-15)

    * Support for references in bracketed spans (pandoc-fignos
      Issue #58).


pandoc-eqnos 1.3.2 (2019-05-15)

    * Enabled equation numbering by section for docx (addresses
      pandoc-fignos Issue #59).


pandoc-eqnos 1.3.1 (2019-05-08)

    * Maintenance release for pandoc 2.7.2.


pandoc-eqnos 1.3.0 (2018-08-09)

    * Require pandoc-xnos>=1.1,<2 to ensure API compatibility
      (pandoc-fignos Issue #46).
    * Gives more helpful error message regarding acceptable
      boolean metadata values (pandoc-fignos Issue #51).


pandoc-eqnos 1.2.0 (2018-01-14)

    * Added `eqnos-capitalise` meta variable.


pandoc-eqnos 1.1.0 (2018-01-13)

    * Added AMS-style equations referencing (Issue #27).


pandoc-eqnos 1.0.0 (2018-01-07)

    * Version number bumped to 1.0.0.


pandoc-eqnos 1.0.0rc3 (2017-12-16)

    * Write equation id in outer span; removed redundant
      "<a name=..." tag (pandoc-eqnos Issue #26).
    * Removed dependency on pandoc-attributes
      (pandoc-fignos Issue #42).


pandoc-eqnos 1.0.0rc1 (2017-11-18)

    * Release candidate updated for pandoc 2.0.


pandoc-eqnos 0.18.1 (2017-09-19)

    * Fixes for beamer (pandoc-fignos Issue #38).


pandoc-eqnos 0.17 (2017-09-19)

    * Bug fix for Word output (Issue #16).


pandoc-eqnos 0.16.1 (2017-09-12)

    * Documentation updates.


pandoc-eqnos 0.16 (2016-12-05)

    * Added numbering equations by section for html (Issue #15).


pandoc-eqnos 0.15 (2016-11-29)

    * For html, write equation numbers in a span and justify them
      right.  Before, the equation numbers were written into
      the equation.  This is a big change in approach.


pandoc-eqnos 0.14.1 (2016-11-17)

    * Python 2.7.3 bug addressed with pandoc-xnos 0.7.
    * Clarified purpose of this filter in README.md.


pandoc-eqnos 0.14 (2016-11-05)

    * Updated for pandoc 1.18 (pandoc-fignos Issue #30)


pandoc-eqnos 0.13 (2016-10-22)

    * Fixed metadata corruption bug (pandoc-fignos Issue #23).
    * Fixed python 2.7 install on Windows (pandoc-fignos Issue #27).
    * Require pandoc-xnos >= 0.5.


pandoc-eqnos 0.12.1 (2016-08-21)

    * Require pandoc-xnos >= 0.4.1


pandoc-eqnos 0.12 (2016-08-21)

    * Removed anchors and ids in output for unreferenceable figures.


pandoc-eqnos 0.11 (2016-05-15)

    * Added clever referencing.
    * Moved library code to pandocxnos project.


pandoc-eqnos 0.10 (2016-04-27)

    * Improved pandoc version detection (Issue #14).


pandoc-eqnos 0.9 (2016-04-19)

    * Allow tagged equations (Issue #9).
    * Improved pandoc version detection (Issue #10).


pandoc-eqnos 0.8 (2016-04-05)

    * Fixed broken installation in Win 10 (Issue #6).
    * Added regression tests.


pandoc-eqnos 0.7.2 (2016-01-22)

    * Fixed AttributeError bug that occurred for links
      emphasized [*like this*](http://example.com/).

    * Fixed TypeError for pandoc < 1.16 when autolink_bare_uris is
      used.


pandoc-eqnos 0.7.1 (2016-01-13)

    * Fixed TypeError bug that occurred with pandoc < 1.16 (Issue #7).
      This bug was inadvertently introduced with pandoc-eqnos 0.7.


pandoc-eqnos 0.7 (2016-01-11)

    * Fixed problems when pandoc's autolink_bare_uris is used.


pandoc-eqnos 0.6 (2016-01-01)


pandoc_eqnos.py
#! /usr/bin/env python

"""pandoc-eqnos: a pandoc filter that inserts equation nos. and refs."""


__version__ = '2.5.0'


# Copyright 2015-2020 Thomas J. Duck.
# All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# OVERVIEW
#
# The basic idea is to scan the document twice in order to:
#
#   1. Insert text for the equation number in each equation.
#      For LaTeX, change to a numbered equation and use \label{...}
#      instead.  The equation labels and associated equation numbers
#      are stored in the global targets tracker.
#
#   2. Replace each reference with an equation number.  For LaTeX,
#      replace with \ref{...} instead.
#
# This is followed by injecting header code as needed for certain output
# formats.

# pylint: disable=invalid-name

import re
import functools
import argparse
import json
import copy
import textwrap
import uuid

from pandocfilters import walk
from pandocfilters import Math, RawInline, Str, Span

import pandocxnos
from pandocxnos import PandocAttributes
from pandocxnos import STRTYPES, STDIN, STDOUT, STDERR
from pandocxnos import elt, check_bool, get_meta
from pandocxnos import repair_refs, process_refs_factory, replace_refs_factory
from pandocxnos import attach_attrs_factory, detach_attrs_factory
from pandocxnos import insert_secnos_factory, delete_secnos_factory
from pandocxnos import version

# Patterns for matching labels and references
LABEL_PATTERN = re.compile(r'(eq:[\w/-]*)')

# Meta variables; may be reset elsewhere
cleveref = False    # Flags that clever references should be used
capitalise = False  # Flags that plusname should be capitalised
plusname = ['eq.', 'eqs.']            # Sets names for mid-sentence references
starname = ['Equation', 'Equations']  # Sets names for refs at sentence start
numbersections = False  # Flags that equations should be numbered by section
secoffset = 0           # Section number offset
eqref = False           # Flags that \eqref should be used
warninglevel = 2        # 0 - no warnings; 1 - some warnings; 2 - all warnings
default_env = 'equation'

# Processing state variables
cursec = None    # Current section
Ntargets = 0  # Number of targets in current section (or document)
targets = {}  # Global targets tracker

# Processing flags
plusname_changed = False          # Flags that the plus name changed
starname_changed = False          # Flags that the star name changed
has_unnumbered_equations = False  # Flags unnumbered equations were found

PANDOCVERSION = None
AttrMath = None


# Actions --------------------------------------------------------------------

# pylint: disable=too-many-branches
def _process_equation(value, fmt):
    """Processes the equation.  Returns a dict containing eq properties."""

    # pylint: disable=global-statement
    global Ntargets                  # Global targets counter
    global cursec                    # Current section
    global has_unnumbered_equations  # Flags that unnumbered eqs were found

    # Initialize the return value
    eq = {'is_unnumbered': False,
          'is_unreferenceable': False,
          'is_tagged': False}

    # Parse the equation
    attrs = eq['attrs'] = PandocAttributes(value[0], 'pandoc')

    # Bail out if the label does not conform to expectations
    if not LABEL_PATTERN.match(attrs.id):
        eq.update({'is_unnumbered':True, 'is_unreferenceable':True})
        return eq

    # Identify unreferenceable equations
    if attrs.id == 'eq:': # Make up a unique description
        attrs.id += str(uuid.uuid4())
        eq['is_unreferenceable'] = True

    # Update the current section number
    if attrs['secno'] != cursec:  # The section number changed
        cursec = attrs['secno']   # Update the global section tracker
        if numbersections:
            Ntargets = 0          # Resets the global targets counter

    # Increment the targets counter
    if 'tag' not in attrs:
        Ntargets += 1

    # Pandoc's --number-sections supports section numbering latex/pdf, html,
    # epub, and docx
    if numbersections:
        # Latex/pdf supports equation numbers by section natively.  For the
        # other formats we must hard-code in equation numbers by section as
        # tags.
        if fmt in ['html', 'html4', 'html5', 'epub', 'epub2', 'epub3', 'docx'] and \
          'tag' not in attrs:
            attrs['tag'] = str(cursec+secoffset) + '.' + str(Ntargets)

    # Save reference information
    eq['is_tagged'] = 'tag' in attrs
    if eq['is_tagged']:   # ... then save the tag
        # Remove any surrounding quotes
        if attrs['tag'][0] == '"' and attrs['tag'][-1] == '"':
            attrs['tag'] = attrs['tag'].strip('"')
        elif attrs['tag'][0] == "'" and attrs['tag'][-1] == "'":
            attrs['tag'] = attrs['tag'].strip("'")
        targets[attrs.id] = pandocxnos.Target(attrs['tag'], cursec,
                                              attrs.id in targets)
    else:
        targets[attrs.id] = pandocxnos.Target(Ntargets, cursec,
                                              attrs.id in targets)

    return eq


def _adjust_equation(fmt, eq, value):
    """Adjusts the equation depending on the output format."""
    attrs = eq['attrs']
    num = targets[attrs.id].num

    if fmt in ['latex', 'beamer']:
        if not eq['is_unreferenceable']:  # Code in the tags
            if eq['is_tagged']:
                value[-1] += r'\tag{%s}\label{%s}' % \
                  (num.replace(' ', r'\ '), attrs.id)
            else:
                value[-1] += r'\label{%s}'%attrs.id
    elif fmt in ('html', 'html4', 'html5', 'epub', 'epub2', 'epub3'):
        pass  # Insert html in _add_markup() instead
    else:  # Hard-code in the number/tag
        if isinstance(num, int):  # Numbered reference
            value[-1] += r'\qquad (%d)' % num
        else:  # Tagged reference
            assert isinstance(num, STRTYPES)
            num = num.replace(' ', r'\ ')
            value[-1] += r'\qquad (%s)' % \
              (num[1:-1] if num.startswith('$') and num.endswith('$') else
               r'\text{%s}' % num)


def _add_markup(fmt, eq, value):
    """Adds markup to the output."""

    attrs = eq['attrs']

    # Context-dependent output
    if eq['is_unnumbered']:  # Unnumbered is also unreferenceable
        ret = None
    elif fmt in ['latex', 'beamer']:
        if 'env' in attrs:
            env = attrs['env']
        else:
            env = default_env
        env, _, arg = env.partition('.')
        ret = RawInline('tex',
                        r'\begin{%s}%s%s\end{%s}'% \
                        (env, '{%s}'%arg if arg else '', value[-1], env))
    elif fmt in ('html', 'html4', 'html5', 'epub', 'epub2', 'epub3') and \
      LABEL_PATTERN.match(attrs.id):
        # Present equation and its number in a span
        num = str(targets[attrs.id].num)
        outer = RawInline('html',
                          '<span%sclass="eqnos">' % \
                            (' ' if eq['is_unreferenceable'] else
                             ' id="%s" '%attrs.id))
        inner = RawInline('html', '<span class="eqnos-number">')
        eqno = Math({"t":"InlineMath"}, '(%s)' % num[1:-1]) \
          if num.startswith('$') and num.endswith('$') \
          else Str('(%s)' % num)
        endtags = RawInline('html', '</span></span>')
        ret = [outer, AttrMath(*value), inner, eqno, endtags]
    elif fmt == 'docx':
        # As per http://officeopenxml.com/WPhyperlink.php
        bookmarkstart = \
          RawInline('openxml',
                    '<w:bookmarkStart w:id="0" w:name="%s"/><w:r><w:t>'
                    %attrs.id)
        bookmarkend = \
          RawInline('openxml',
                    '</w:t></w:r><w:bookmarkEnd w:id="0"/>')
        ret = [bookmarkstart, AttrMath(*value), bookmarkend]
    else:
        ret = None
    return ret


def process_equations(key, value, fmt, meta):  # pylint: disable=unused-argument
    """Processes the attributed equations."""

    # Process attributed equations and add markup
    if key == 'Math' and len(value) == 3:
        eq = _process_equation(value, fmt)
        if eq['attrs'].id:
            _adjust_equation(fmt, eq, value)
        return _add_markup(fmt, eq, value)

    return None


# TeX blocks -----------------------------------------------------------------

# Define some tex to number equations by section
NUMBER_BY_SECTION_TEX = r"""
%% pandoc-eqnos: number equations by section
\numberwithin{equation}{section}
"""

# Section number offset
SECOFFSET_TEX = r"""
%% pandoc-eqnos: section number offset
\setcounter{section}{%s}
"""

# Define some tex to disable brackets around cleveref numbers
DISABLE_CLEVEREF_BRACKETS_TEX = r"""
%% pandoc-eqnos: disable brackets around cleveref numbers
\creflabelformat{equation}{#2#1#3}
"""


# Html blocks ----------------------------------------------------------------

# Equation css
# Valid XHTML 1.0 Transitional (called html4 in pandoc) requires type="text/css"
# See also https://github.com/tomduck/pandoc-eqnos/issues/50
EQUATION_STYLE_HTML = """
<!-- pandoc-eqnos: equation style -->
<style%s>
  .eqnos { display: inline-block; position: relative; width: 100%%; }
  .eqnos br { display: none; }
  .eqnos-number { position: absolute; right: 0em; top: 50%%; line-height: 0; }
</style>
"""


# Main program ---------------------------------------------------------------

# pylint: disable=too-many-statements
def process(meta):
    """Saves metadata fields in global variables and returns a few
    computed fields."""

    # pylint: disable=global-statement
    global cleveref    # Flags that clever references should be used
    global capitalise  # Flags that plusname should be capitalised
    global plusname    # Sets names for mid-sentence references
    global starname    # Sets names for references at sentence start
    global numbersections  # Flags that sections should be numbered by section
    global secoffset       # Section number offset
    global warninglevel    # 0 - no warnings; 1 - some; 2 - all
    global plusname_changed  # Flags that the plus name changed
    global starname_changed  # Flags that the star name changed
    global eqref             # Flags that \eqref should be used
    global default_env       # Default equations environment

    # Read in the metadata fields and do some checking

    for name in ['eqnos-warning-level', 'xnos-warning-level']:
        if name in meta:
            warninglevel = int(get_meta(meta, name))
            pandocxnos.set_warning_level(warninglevel)
            break

    metanames = ['eqnos-warning-level', 'xnos-warning-level',
                 'eqnos-cleveref', 'xnos-cleveref',
                 'xnos-capitalise', 'xnos-capitalize',
                 'xnos-caption-separator', # Used by pandoc-fignos/tablenos
                 'eqnos-plus-name', 'eqnos-star-name',
                 'eqnos-number-by-section', 'xnos-number-by-section',
                 'xnos-number-offset',
                 'eqnos-eqref',
                 'eqnos-default-env']

    if warninglevel:
        for name in meta:
            if (name.startswith('eqnos') or name.startswith('xnos')) and \
              name not in metanames:
                msg = textwrap.dedent("""
                          pandoc-eqnos: unknown meta variable "%s"\n
                      """ % name)
                STDERR.write(msg)

    for name in ['eqnos-cleveref', 'xnos-cleveref']:
        # 'xnos-cleveref' enables cleveref in all 3 of fignos/eqnos/tablenos
        if name in meta:
            cleveref = check_bool(get_meta(meta, name))
            break

    for name in ['xnos-capitalise', 'xnos-capitalize']:
        # 'xnos-capitalise' enables capitalise in all 3 of
        # fignos/eqnos/tablenos.  Since this uses an option in the caption
        # package, it is not possible to select between the three (use
        # 'eqnos-plus-name' instead.  'xnos-capitalize' is an alternative
        # spelling
        if name in meta:
            capitalise = check_bool(get_meta(meta, name))
            break

    if 'eqnos-plus-name' in meta:
        tmp = get_meta(meta, 'eqnos-plus-name')
        old_plusname = copy.deepcopy(plusname)
        if isinstance(tmp, list):  # The singular and plural forms were given
            plusname = tmp
        else:  # Only the singular form was given
            plusname[0] = tmp
        plusname_changed = plusname != old_plusname
        assert len(plusname) == 2
        for name in plusname:
            assert isinstance(name, STRTYPES)
        if plusname_changed:
            starname = [name.title() for name in plusname]

    if 'eqnos-star-name' in meta:
        tmp = get_meta(meta, 'eqnos-star-name')
        old_starname = copy.deepcopy(starname)
        if isinstance(tmp, list):
            starname = tmp
        else:
            starname[0] = tmp
        starname_changed = starname != old_starname
        assert len(starname) == 2
        for name in starname:
            assert isinstance(name, STRTYPES)

    for name in ['eqnos-number-by-section', 'xnos-number-by-section']:
        if name in meta:
            numbersections = check_bool(get_meta(meta, name))
            break

    if 'xnos-number-offset' in meta:
        secoffset = int(get_meta(meta, 'xnos-number-offset'))

    if 'eqnos-eqref' in meta:
        eqref = check_bool(get_meta(meta, 'eqnos-eqref'))
        # Note: Eqref and cleveref are mutually exclusive.  If both are
        # enabled, then cleveref will be used but with bracketed equation
        # numbers.

    if 'eqnos-default-env' in meta:
        default_env = get_meta(meta, 'eqnos-default-env')

def add_tex(meta):
    """Adds tex to the meta data."""

    warnings = warninglevel == 2 and targets and \
      (pandocxnos.cleveref_required() or
       plusname_changed or starname_changed or numbersections or secoffset)
    if warnings:
        msg = textwrap.dedent("""\
                  pandoc-eqnos: Wrote the following blocks to
                  header-includes.  If you use pandoc's
                  --include-in-header option then you will need to
                  manually include these yourself.
              """)
        STDERR.write('\n')
        STDERR.write(textwrap.fill(msg))
        STDERR.write('\n')

    # Update the header-includes metadata.  Pandoc's
    # --include-in-header option will override anything we do here.  This
    # is a known issue and is owing to a design decision in pandoc.
    # See https://github.com/jgm/pandoc/issues/3139.

    if pandocxnos.cleveref_required() and targets:
        tex = """
            %%%% pandoc-eqnos: required package
            \\usepackage%s{cleveref}
        """ % ('[capitalise]' if capitalise else '')
        pandocxnos.add_to_header_includes(
            meta, 'tex', tex,
            regex=r'\\usepackage(\[[\w\s,]*\])?\{cleveref\}')

        if not eqref:
            pandocxnos.add_to_header_includes(
                meta, 'tex', DISABLE_CLEVEREF_BRACKETS_TEX)

    if plusname_changed and targets:
        tex = """
            %%%% pandoc-eqnos: change cref names
            \\crefname{equation}{%s}{%s}
        """ % (plusname[0], plusname[1])
        pandocxnos.add_to_header_includes(meta, 'tex', tex)

    if starname_changed and targets:
        tex = """
            %%%% pandoc-eqnos: change Cref names
            \\Crefname{equation}{%s}{%s}
        """ % (starname[0], starname[1])
        pandocxnos.add_to_header_includes(meta, 'tex', tex)

    if numbersections and targets:
        pandocxnos.add_to_header_includes(meta, 'tex', NUMBER_BY_SECTION_TEX)

    if secoffset and targets:
        pandocxnos.add_to_header_includes(
            meta, 'tex', SECOFFSET_TEX % secoffset,
            regex=r'\\setcounter\{section\}')

    if warnings:
        STDERR.write('\n')

def add_html(meta, fmt):
    """Adds html to the meta data."""

    warnings = warninglevel == 2 and targets

    if warnings:
        msg = textwrap.dedent("""\
                  pandoc-eqnos: Wrote the following blocks to
                  header-includes.  If you use pandoc's
                  --include-in-header option then you will need to
                  manually include these yourself.
              """)
        STDERR.write('\n')
        STDERR.write(textwrap.fill(msg))
        STDERR.write('\n')

    # Update the header-includes metadata.  Pandoc's
    # --include-in-header option will override anything we do here.  This
    # is a known issue and is owing to a design decision in pandoc.
    # See https://github.com/jgm/pandoc/issues/3139.

    if targets:
        cond = fmt == 'html4' or \
          (fmt == 'html' and version(PANDOCVERSION) < version('2.0'))
        attr = ' type="text/css"' if cond else ''
        pandocxnos.add_to_header_includes(meta, 'html',
                                          EQUATION_STYLE_HTML%attr)

# pylint: disable=too-many-locals, unused-argument
def main(stdin=STDIN, stdout=STDOUT, stderr=STDERR):
    """Filters the document AST."""

    # pylint: disable=global-statement
    global PANDOCVERSION
    global AttrMath

    # Read the command-line arguments
    parser = argparse.ArgumentParser(\
      description='Pandoc equations numbers filter.')
    parser.add_argument(\
      '--version', action='version',
      version='%(prog)s {version}'.format(version=__version__))
    parser.add_argument('fmt')
    parser.add_argument('--pandocversion', help='The pandoc version.')
    args = parser.parse_args()

    # Get the output format and document
    fmt = args.fmt
    doc = json.loads(stdin.read())

    # Initialize pandocxnos
    PANDOCVERSION = pandocxnos.init(args.pandocversion, doc)

    # Element primitives
    AttrMath = elt('Math', 3)

    # Chop up the doc
    meta = doc['meta'] if version(PANDOCVERSION) >= version('1.18') \
      else doc[0]['unMeta']
    blocks = doc['blocks'] if version(PANDOCVERSION) >= version('1.18') \
      else doc[1:]

    # Process the metadata variables
    process(meta)

    # First pass
    attach_attrs_math = attach_attrs_factory(Math, allow_space=True)
    detach_attrs_math = detach_attrs_factory(Math)
    insert_secnos = insert_secnos_factory(Math)
    delete_secnos = delete_secnos_factory(Math)
    altered = functools.reduce(lambda x, action: walk(x, action, fmt, meta),
                               [attach_attrs_math, insert_secnos,
                                process_equations, delete_secnos,
                                detach_attrs_math], blocks)

    # Second pass
    process_refs = process_refs_factory(LABEL_PATTERN, targets.keys())
    replace_refs = replace_refs_factory(targets,
                                        cleveref, eqref,
                                        plusname if not capitalise or \
                                        plusname_changed else
                                        [name.title() for name in plusname],
                                        starname)
    attach_attrs_span = attach_attrs_factory(Span, replace=True)
    altered = functools.reduce(lambda x, action: walk(x, action, fmt, meta),
                               [repair_refs, process_refs, replace_refs,
                                attach_attrs_span],
                               altered)

    if fmt in ['latex', 'beamer']:
        add_tex(meta)
    elif fmt in ['html', 'html4', 'html5', 'epub', 'epub2', 'epub3']:
        add_html(meta, fmt)

    # Update the doc
    if version(PANDOCVERSION) >= version('1.18'):
        doc['blocks'] = altered
    else:
        doc = doc[:1] + altered

    # Dump the results
    json.dump(doc, stdout)

    # Flush stdout
    stdout.flush()

if __name__ == '__main__':
    main()
